homeassistant/components/husqvarna_automower/coordinator.py

```diff
@@ -35,13 +41,38 @@ async def _async_update_data(self) -> dict[str, MowerAttributes]:
             await self.api.connect()
             self.api.register_data_callback(self.callback)
             self.ws_connected = True
-        return await self.api.get_status()
-
-    async def shutdown(self, *_: Any) -> None:
-        """Close resources."""
-        await self.api.close()
+        try:
+            return await self.api.get_status()
+        except ApiException as err:
+            raise UpdateFailed(err) from err
 
     @callback
     def callback(self, ws_data: dict[str, MowerAttributes]) -> None:
         """Process websocket callbacks and write them to the DataUpdateCoordinator."""
         self.async_set_updated_data(ws_data)
+
+    async def client_listen(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        automower_client: AutomowerSession,
+        reconnect_time: int = 2,
+    ) -> None:
+        """Listen with the client."""
+        try:
+            websocket = await automower_client.auth.websocket_connect()
+            await automower_client.start_listening(websocket)
```

Comment:
I don't think the library should expose the websocket directly. Store the websocket on the client in the connect method after successful connection. Let the start listening method use the instance attribute that holds the websocket.

Reply:
Maybe use the ready event to timeout the listen call if we don't get the ready message after some time? Then disconnect and try to reconnect. Does the ready event fail to come anytime?

Reply:
> Maybe use the ready event to timeout the listen call if we don't get the ready message after some time? Then disconnect and try to reconnect. Does the ready event fail to come anytime?

So far the ready event only didn't come, when there was no internet connection. But in regular conditions it always works.

Reply:
Ok.

Reply:
So it's okay to not use the ready event?

Reply:
Yeah, you know the Husqvarna api best. Do what you think works best.